### PR TITLE
FMWK-199 Refactor implementation of paginated queries

### PR DESCRIPTION
**Pull request by agrgr**

### CODE REVIEW

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 1]
```diff
@@ -1177,6 +1177,12 @@ private <T> Flux<T> findUsingQueryWithPostProcessing(String setName, Class<T> ta
         return results;
     }
 
+    public <T> Flux<T> findUsingQueryWithoutPostProcessing(Class<?> entityClass, Class<T> targetClass, Query query) {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java:1180]: Add `@Override`

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeOperations.java [hunk 2]
```diff
@@ -868,6 +868,8 @@ <T, S> Flux<?> findByIdsUsingQuery(Collection<?> ids, Class<T> entityClass, Clas
      */
     <T, S> Flux<S> findInRange(long offset, long limit, Sort sort, Class<T> entityClass, Class<S> targetClass);
 
+    <T> Flux<T> findUsingQueryWithoutPostProcessing(Class<?> entityClass, Class<T> targetClass, Query query);
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeOperations.java:871]: Javadoc is required

File: src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java [hunk 3]
```diff
@@ -896,6 +896,8 @@ <T, S> List<?> findByIdsUsingQuery(Collection<?> ids, Class<T> entityClass, Clas
      */
     <T> Stream<T> findInRange(long offset, long limit, Sort sort, Class<T> targetClass, String setName);
 
+    <T> Stream<T> findUsingQueryWithoutPostProcessing(Class<?> entityClass, Class<T> targetClass, Query query);
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java:899]: Javadoc is required

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 4]
```diff
@@ -944,6 +944,12 @@ private <T> Stream<T> findUsingQueryWithPostProcessing(String setName, Class<T>
         return applyPostProcessingOnResults(results, query);
     }
 
+    public <T> Stream<T> findUsingQueryWithoutPostProcessing(Class<?> entityClass, Class<T> targetClass, Query query) {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:947]: Add `@Override`

File: src/main/java/org/springframework/data/aerospike/query/QueryEngine.java [hunk 5]
```diff
@@ -54,6 +54,7 @@ public class QueryEngine {
      * scans, set this property to true.
      */
     private boolean scansEnabled = false;
+    private int queryMaxRecords = 100000;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/QueryEngine.java:57]: Should be taken from `AerospikeDataSettings`?

File: src/main/java/org/springframework/data/aerospike/query/QueryEngine.java [hunk 6]
```diff
@@ -128,6 +130,11 @@ public void setScansEnabled(boolean scansEnabled) {
         this.scansEnabled = scansEnabled;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/QueryEngine.java:130]: `@Setter`?

File: src/main/java/org/springframework/data/aerospike/query/QueryEngine.java [hunk 7]
```diff
@@ -128,6 +130,11 @@ public void setScansEnabled(boolean scansEnabled) {
...
     }
 
+    public void setQueryMaxRecords(int queryMaxRecords) {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/QueryEngine.java:133]: `@Setter`?

File: src/main/java/org/springframework/data/aerospike/query/ReactorQueryEngine.java [hunk 8]
```diff
@@ -115,4 +117,8 @@ private Mono<KeyRecord> getRecord(Policy policy, Key key, String[] binNames) {
     public void setScansEnabled(boolean scansEnabled) {
         this.scansEnabled = scansEnabled;
     }
+
+    public void setQueryMaxRecords(int queryMaxRecords) {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/ReactorQueryEngine.java:121]: `@Setter`?

File: src/main/java/org/springframework/data/aerospike/query/ReactorQueryEngine.java [hunk 9]
```diff
@@ -115,4 +117,8 @@ private Mono<KeyRecord> getRecord(Policy policy, Key key, String[] binNames) {
     public void setScansEnabled(boolean scansEnabled) {
         this.scansEnabled = scansEnabled;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/ReactorQueryEngine.java:118]: `@Setter`?

File: src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java [hunk 10]
```diff
@@ -896,6 +896,18 @@ <T, S> List<?> findByIdsUsingQuery(Collection<?> ids, Class<T> entityClass, Clas
...
+     * Find documents in the given entityClass's set using a query and map them to the given target class type. If the
+     * query has pagination and/or sorting, post-processing must be applied separately.
+     *
+     * @param entityClass The class to extract the Aerospike set from. Must not be {@literal null}.
+     * @param targetClass The class to map the document to.
+     * @param query       The {@link Query} to filter results.
+     * @return A Stream of all matching documents regardless of pagination/sorting, returned documents will be mapped to
+     * targetClass's type.
+     */
+    <T> Stream<T> findUsingQueryWithoutPostProcessing(Class<?> entityClass, Class<T> targetClass, Query query);
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java:909]: Why does entityClass is wrapped with Class<?>, usually when we have entity and target we use Class<T> for entityClass and Class<S> for targetClass

File: src/main/java/org/springframework/data/aerospike/repository/query/ReactiveAerospikePartTreeQuery.java [hunk 11]
```diff
@@ -35,14 +38,14 @@
  */
 public class ReactiveAerospikePartTreeQuery extends BaseAerospikePartTreeQuery {
 
-    private final ReactiveAerospikeOperations operations;
+    private final ReactiveAerospikeTemplate template;
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/repository/query/ReactiveAerospikePartTreeQuery.java:41]: Same (why template?)

File: src/test/java/org/springframework/data/aerospike/repository/PersonRepositoryQueryTests.java [hunk 12]
```diff
@@ -1322,8 +1322,8 @@ public void findPersonsByQuery() {
             .setFilterOperation(FilterOperation.EQ)
             .setValue1(Value.get(49))
             .build();
-        result = repository.findUsingQuery(new Query(ageEq49));
-        assertThat(result).containsOnly(carter);
+//        result = repository.findUsingQuery(new Query(ageEq49));
```
Comment (roimenashe) [inline src/test/java/org/springframework/data/aerospike/repository/PersonRepositoryQueryTests.java:1325]: Why commenting assertions? if its not relevant anymore remove

File: src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java [hunk 13]
```diff
@@ -896,6 +902,18 @@ <T, S> List<?> findByIdsUsingQuery(Collection<?> ids, Class<T> entityClass, Clas
...
+     * Find documents in the given entityClass's set using a query and map them to the given target class type. If the
+     * query has pagination and/or sorting, post-processing must be applied separately.
+     *
+     * @param entityClass The class to extract the Aerospike set from. Must not be {@literal null}.
+     * @param targetClass The class to map the document to.
+     * @param query       The {@link Query} to filter results.
+     * @return A Stream of all matching documents regardless of pagination/sorting, returned documents will be mapped to
+     * targetClass's type.
+     */
+    <T, S> Stream<T> findUsingQueryWithoutPostProcessing(Class<S> entityClass, Class<T> targetClass, Query query);
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java]: Please switch between S and T (T for entity and S for TargetClass) in both interfaces (reactive and non-reactive) and both implementations, total of 4 places) as we talked about.

**Review by roimenashe — APPROVED** [review]

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 14]
```diff
@@ -1044,15 +1057,34 @@ public <T> long count(Query query, Class<T> entityClass) {
...
+    private Stream<KeyRecord> countRecordsUsingQuery(String setName, Query query) {
         Assert.notNull(query, "Query must not be null!");
         Assert.notNull(setName, "Set name must not be null!");
 
-        return findRecordsUsingQuery(setName, null, query);
+        Qualifier qualifier = queryCriteriaIsNotNull(query) ? query.getQualifier() : null;
+        if (qualifier != null) {
+            Qualifier idQualifier = getOneIdQualifier(qualifier);
+            if (idQualifier != null) {
+                // a special flow if there is id given
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:1072]: Sounds grammatically wrong, maybe: 
// Dedicate flow when id is given

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 15]
```diff
@@ -974,14 +974,23 @@ public Mono<Long> count(Query query, String setName) {
...
+    private Flux<KeyRecord> countRecordsUsingQuery(String setName, Query query) {
         Assert.notNull(query, "Query must not be null!");
         Assert.notNull(setName, "Set name must not be null!");
 
-        return findRecordsUsingQuery(setName, null, query);
+        Qualifier qualifier = queryCriteriaIsNotNull(query) ? query.getQualifier() : null;
+        if (qualifier != null) {
+            Qualifier idQualifier = getOneIdQualifier(qualifier);
+            if (idQualifier != null) {
+                // a special flow if there is id given
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java:988]: Same

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 14 above)
Comment (agrgr) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: Agree, updated to a similar phrase

**Review by reugn — APPROVED** [review]

**Pull request merged**